### PR TITLE
Feature Infrastructure Service

### PR DESCRIPTION
**Pull request by gerardcl**
Proposal to fix https://github.com/opendevstack/ods-quickstarters/issues/631

For now the proposal is to have stage `odsComponentStageInfrastructureAWS` that uses `InfrastructureService` which wraps the `Makefile` make executions (similar approach as other tooling stages like Aqua). The config param expected are the `envPath`, which the default is as per current AWS quickstarter envs path `./environments`, and the `resourceName` used for tracking timings.

In future we could consider naming it `odsComponentStageInfrastructure` and pass the provider as a config param, so any quickstarter specialized on a provider (e.g.: AWS, Azure, GCP,...) would just need the right environments and Makefile in place.

The current approach will also let users have in the same component used for business logic (e.g.: nodejs backend) to leveraging the related resources in their infrastructure (e.g.: RDS or S3), keeping a nice micro-service config approach.

This is how the current QS Jenkinsfile would look like now:

```
/* generated jenkins file used for building and deploying AWS-infrastructure in projects */

@Library('ods-jenkins-shared-library@4.x') _

odsComponentPipeline(
  imageStreamTag: 'ods/jenkins-agent-terraform:4.x',
  branchToEnvironmentMapping: [
    'master': 'dev',
  ]
) { context ->
    odsComponentStageInfrastructureAWS(context)
}
```

TODO:
- [x] Add docs
- [x] Add tests
- [x] Changelog

@metmajer shall I move the target issue to this repo?

FYI @metmajer @michaelsauter @tbugfinder @nichtraunzer @clemensutschig let me know your thoughts! 
Question: would you agree and see no issues on adding (sub)stages within curren proposed stage (so to expose to jenkins specific stages for testing, build,... all called inside the main IaC stage)? Basically, with the new approach we only get one stage one:

Screenshot 2021-10-18 at 11 47 36

And the previous one (which we could add back as suggested, with sub-stages) looks like:

Screenshot 2021-10-18 at 11 56 06

### CODE REVIEW

**Comment by nichtraunzer** [comment]
@gerardcl after a discussion with @metmajer I agree with him and lets skip the first todo and lets focus on the core capabilities of  the QS first. 
Regarding PreDeploymentTesting i would like to go a step further and skip this step at all. Why doing it when it already has been executed multiple times by developers.
I propose to focus on the AWS implementation for now.

**Comment by gerardcl** [comment]
@nichtraunzer @metmajer  good that you met and aligned 👍 LGTM! 🚀 
let's then meet any time soon and continue from here :)

**Comment by nichtraunzer** [comment]
hi! happy new year! :) @gerardcl and I are retaking this topic and we are noting down here latest agreements:

TODOs:
- docs update
- tests addition
- changelog update
- change name of smoke stage
- expose all substages as stages to jenkins
- do sync changes in a PR in ods-quickstarters to use this new service (inf-terraform-aws)

Next steps will be discussed afterwards.
Ideas -> align this changes with the inf-terraform-azure

@metmajer @tbugfinder FYI

**Comment by metmajer** [comment]
@gerardcl @nichtraunzer it would be great if we could close this down soon. Any major topics open? This would enable the Azure Quickstarter to be built in a similar fashion without exposing the entire logic in the component pipeline. Thanks for the efforts!

**Comment by nichtraunzer** [comment]
> @gerardcl @nichtraunzer it would be great if we could close this down soon. Any major topics open? This would enable the Azure Quickstarter to be built in a similar fashion without exposing the entire logic in the component pipeline. Thanks for the efforts!
@metmajer we plan to release the first draft by next week. Still doing some testing and documentation.

**Review by nichtraunzer — APPROVED** [review]
nice job

**Comment by nichtraunzer** [comment]
We think that the current version is ready for a review @metmajer @tbugfinder @michaelsauter @kuebler 
The current proposal has shift to  contain the initial steps to support multiple cloud providers - see 

https://github.com/opendevstack/ods-quickstarters/pull/749/files#diff-a58bdbe85174e45a7fbe5363b108c78b2936fca8e38562836482589084f56c9bR11

File: docs/modules/jenkins-shared-library/partials/odsComponentStageInfrastructure.adoc
Comment (metmajer) [inline docs/modules/jenkins-shared-library/partials/odsComponentStageInfrastructure.adoc:24]: @nichtraunzer what exactly would the QS do, if a user names one or several branch or branches? Can you be a bit more concrete?

File: src/org/ods/component/InfrastructureStage.groovy
Comment (metmajer) [reply src/org/ods/component/InfrastructureStage.groovy:41]: Seems like the code portion is still part of the final code base. Do we really need a boolean flag `stackDeploy`? Will the code do anything meaningful if it's false?

**Review by metmajer — APPROVED** [review]
Great work overall, @nichtraunzer @gerardcl @tbugfinder. Some minor improvements asked for in the questions.

File: docs/modules/jenkins-shared-library/partials/odsComponentStageInfrastructure.adoc
Comment (nichtraunzer) [reply docs/modules/jenkins-shared-library/partials/odsComponentStageInfrastructure.adoc:24]: that file is auto generated ... all the ods stages have that.

Comment (metmajer) [reply docs/modules/jenkins-shared-library/partials/odsComponentStageInfrastructure.adoc:24]: I am thinking of creating more immediately understandable documentation for technical end-users. Might be good enough for now, some use case example wouldn't hurt. We can keep this as-is.

**Comment by gerardcl** [comment]
hi! so this is now ready to get a final review from all involved in this. Please, note that once this is merged (to master) we will need (correct me if I am wrong) to port this changes to ODS 4.x too.

**Comment by nichtraunzer** [comment]
Thanks @gerardcl for the great support

File: docs/modules/jenkins-shared-library/partials/odsComponentStageInfrastructure.adoc
Comment (michaelsauter) [inline docs/modules/jenkins-shared-library/partials/odsComponentStageInfrastructure.adoc:6]: ?

**Review by michaelsauter — COMMENTED** [review]
I only had a brief glance, but overall this looks fine to me.

File: docs/modules/jenkins-shared-library/partials/odsComponentStageInfrastructure.adoc
Comment (nichtraunzer) [reply docs/modules/jenkins-shared-library/partials/odsComponentStageInfrastructure.adoc]: old comments, have been removed

**Review by michaelsauter — APPROVED** [review]

**Comment by nichtraunzer** [comment]
Thanks @gerardcl and get well soon!

**Comment by metmajer** [comment]
@nichtraunzer can we please have the same changes in the 4.x branch via a dedicated PR?

**Comment by nichtraunzer** [comment]
@metmajer I would like to run a complete end 2 end test together with the EDP first before moving to 4.x

**Pull request merged**
